### ISSUE TITLE
What's the design philosophy of Endpoint?

### ISSUE THREAD

**Issue by mutedstack** [ericls/huobi#3]
Normally, we will use a method with position or default params to represent a API function such as [PyGithub](https://github.com/PyGithub/PyGithub/blob/master/github/Issue.py) and [python-binance](https://github.com/sammchardy/python-binance/blob/master/binance/client.py), but here is a instance of `Endpoint` that using params of `Endpoint` to represent different APIs.

Here are my options:

## Advantages

Reuse the codes of `Endpoint`, such as check params, handling response and etc.

## Disadvantages

* Can't use auto-completion as mentioned in #2
* The `Endpoint` should consider all the situations which may make it more complicated

**Comment by ericls** [comment]
Very good question! 

This `Endpoint` approach is something I've never done before, so when writing this small library, I chose the more adventurous path to make the journey more exciting for myself. I actually didn't think of any design patterns or philosophy when I started.

I agree that this `Endpoint` approach does tend to have a high complexity especially if the behavior of the API changes a lot. In fact, the complexity of the `__get__` method is already unnecessarily high due to bad planning. But the advantage is making it "declarative" to add/change apis when there are small updates from upstream, even people new to Python would be able to contribute, because they only need to describe an endpoint instead of writing code to control how that's executed. 

I won't be reluctant to completely rewrite the library (keep the same user faced api, of course) if it became clear to me one approach is better than the other. 

And I think there are ways to combine the advantages of both approaches, at least we should solve the auto completion issue (#2).  We can discuss this later.
